### PR TITLE
Fix  join to subqueries rewriter null pointer

### PR DESCRIPTION
**Pull request by zzsmdfj**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

A null pointer will be generated when select if  as from ‘three table join’ , For example, the SQL:

`
SELECT (if(a.test == 'a', b.test, c.test)) as `a.test` FROM
    (SELECT 1 AS id, 'a' AS test) a
    LEFT JOIN (SELECT 1 AS id, 'b' AS test) b ON b.id = a.id
    LEFT JOIN (SELECT 1 AS id, 'c' AS test) c ON c.id = a.id;
`

### CODE REVIEW

File: src/Interpreters/JoinToSubqueryTransformVisitor.cpp [hunk 1]
```diff
@@ -546,10 +546,13 @@ std::vector<TableNeededColumns> normalizeColumnNamesExtractNeeded(
...
-                    bool alias_equals_column_name = alias_ident->getColumnNameWithoutAlias() == ident->getColumnNameWithoutAlias();
-                    if (!alias_equals_column_name)
-                        throw Exception("Alias clashes with qualified column '" + ident->name() + "'", ErrorCodes::AMBIGUOUS_COLUMN_NAME);
+                    if ( alias_ident->as<ASTIdentifier>())
+                    {
+                        alias_ident->as<ASTIdentifier>()->restoreTable();
+                        bool alias_equals_column_name = alias_ident->getColumnNameWithoutAlias() == ident->getColumnNameWithoutAlias();
+                        if (!alias_equals_column_name)
+                            throw Exception("Alias clashes with qualified column '" + ident->name() + "'", ErrorCodes::AMBIGUOUS_COLUMN_NAME);
+                    }
```
Comment (evillique) [inline src/Interpreters/JoinToSubqueryTransformVisitor.cpp:555]:
```suggestion
                    if (auto * alias_ident_typed = alias_ident->as<ASTIdentifier>())
                    {
                        alias_ident_typed->restoreTable();
                        bool alias_equals_column_name = alias_ident->getColumnNameWithoutAlias() == ident->getColumnNameWithoutAlias();
                        if (!alias_equals_column_name)
                            throw Exception("Alias clashes with qualified column '" + ident->name() + "'", ErrorCodes::AMBIGUOUS_COLUMN_NAME);
                    }
```

**Comment by evillique** [comment]
Style check failed here because of the space after opening the square bracket, and other checks are not run because of that: https://github.com/ClickHouse/ClickHouse/blob/65e1eba1aff530b4e25a69e0685c0cefcb25dded/src/Interpreters/JoinToSubqueryTransformVisitor.cpp#L549

**Comment by zzsmdfj** [comment]
> Style check failed here because of the space after opening the square bracket, and other checks are not run because of that:
> 
> https://github.com/ClickHouse/ClickHouse/blob/65e1eba1aff530b4e25a69e0685c0cefcb25dded/src/Interpreters/JoinToSubqueryTransformVisitor.cpp#L549



> Style check failed here because of the space after opening the square bracket, and other checks are not run because of that:
> 
> https://github.com/ClickHouse/ClickHouse/blob/65e1eba1aff530b4e25a69e0685c0cefcb25dded/src/Interpreters/JoinToSubqueryTransformVisitor.cpp#L549

done!   I thought Suggested change could be applied directly to the submission.

**Review by evillique — APPROVED** [review]

**Comment by evillique** [comment]
Test failures are unrelated.

**Pull request merged**